hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java

```diff
@@ -0,0 +1,530 @@
+
+              try (TableIterator<String, ? extends Table.KeyValue<String,
+                  OmKeyInfo>> deletedDirIterator = snapDeletedDirTable
+                  .iterator(dbBucketKeyForDir)) {
+
+                while (deletedDirIterator.hasNext()) {
+                  Table.KeyValue<String, OmKeyInfo> deletedDirInfo =
+                      deletedDirIterator.next();
+
+                  // For each deleted directory we do an in-memory DFS and
+                  // do a deep clean and exclusive size calculation.
+                  iterateDirectoryTree(deletedDirInfo, volumeId, bucketInfo,
+                      previousSnapshot, previousToPrevSnapshot,
+                      currOmSnapshot, previousKeyTable, prevRenamedTable,
+                      previousToPrevKeyTable, dbBucketKeyForDir);
+                }
+                updateDeepCleanSnapshotDir(currSnapInfo.getTableKey());
+                if (previousSnapshot != null) {
+                  updateExclusiveSize(previousSnapshot.getTableKey());
+                }
+              }
+            }
+          } finally {
+            IOUtils.closeQuietly(rcPrevOmSnapshot, rcPrevToPrevOmSnapshot);
+          }
+        }
+      } catch (IOException ex) {
+        LOG.error("Error while running directory deep clean on snapshots." +
+            " Will retry at next run.", ex);
+      }
+      return BackgroundTaskResult.EmptyTaskResult.newResult();
+    }
+  }
+
+  @SuppressWarnings("checkstyle:ParameterNumber")
+  private void iterateDirectoryTree(
+      Table.KeyValue<String, OmKeyInfo> deletedDirInfo, long volumeId,
+      OmBucketInfo bucketInfo,
+      SnapshotInfo previousSnapshot,
+      SnapshotInfo previousToPrevSnapshot,
+      OmSnapshot currOmSnapshot,
+      Table<String, OmKeyInfo> previousKeyTable,
+      Table<String, String> prevRenamedTable,
+      Table<String, OmKeyInfo> previousToPrevKeyTable,
+      String dbBucketKeyForDir) throws IOException {
+
+    Table<String, OmDirectoryInfo> snapDirTable =
+        currOmSnapshot.getMetadataManager().getDirectoryTable();
+    Table<String, String> snapRenamedTable =
+        currOmSnapshot.getMetadataManager().getSnapshotRenamedTable();
+
+    Stack<StackNode> stackNodes = new Stack<>();
+    OmDirectoryInfo omDeletedDirectoryInfo =
+        getDirectoryInfo(deletedDirInfo.getValue());
+    String dirPathDbKey = currOmSnapshot.getMetadataManager()
+        .getOzonePathKey(volumeId, bucketInfo.getObjectID(),
+            omDeletedDirectoryInfo);
+    // Stack Init
+    StackNode topLevelDir = new StackNode();
+    topLevelDir.setDirKey(dirPathDbKey);
+    topLevelDir.setDirValue(omDeletedDirectoryInfo);
+    stackNodes.push(topLevelDir);
+
+    try (
+        TableIterator<String, ? extends Table.KeyValue<String, OmDirectoryInfo>>
+            directoryIterator = snapDirTable.iterator(dbBucketKeyForDir)) {
+
+      while (!stackNodes.isEmpty()) {
+        StackNode stackTop = stackNodes.pop();
```

Comment:
```suggestion
        StackNode stackTop = stackNodes.peek();
        String seekDirInDB;
        // First process all the files in the current directory
        // and then do a DFS for directory.
        if (StringUtils.isEmpty(stackTop.getSubDirSeek())) {
          processFilesUnderDir(previousSnapshot,
              previousToPrevSnapshot,
              volumeId,
              bucketInfo,
              stackTop.getDirValue(),
              currOmSnapshot.getMetadataManager(),
              snapRenamedTable,
              previousKeyTable,
              prevRenamedTable,
              previousToPrevKeyTable);
          seekDirInDB = currOmSnapshot.getMetadataManager()
              .getOzonePathKey(volumeId, bucketInfo.getObjectID(),
                  stackTop.getDirValue().getObjectID(), "");
          stackTop.setSubDirSeek(seekDirInDB);
        } else {
          directoryIterator.seek(stackTop.getSubDirSeek() + "\0");
          if (directoryIterator.hasNext()) {
            Table.KeyValue<String, OmDirectoryInfo> deletedSubDirInfo =
                directoryIterator.next();
            String deletedSubDirKey = deletedSubDirInfo.getKey();

            String prefixCheck = currOmSnapshot.getMetadataManager()
                .getOzoneDeletePathDirKey(stackTop.getSubDirSeek());
            // Exit if it is out of the sub dir prefix scope.
            if (!deletedSubDirKey.startsWith(prefixCheck)) {
              // Add exit condition.
              stackNodes.pop();
            } else {
              stackTop.setSubDirSeek(deletedSubDirKey);
              StackNode nextSubDir = new StackNode();
              nextSubDir.setDirKey(deletedSubDirInfo.getKey());
              nextSubDir.setDirValue(deletedSubDirInfo.getValue());
              stackNodes.push(nextSubDir);
            }
          } else {
            stackNodes.pop();
          }
        }
```

Reply:
Does this implementation look simpler? I have added "\0" at the end of subDir seek, so it will be the string just greater than the seeked string and next string will be iterated. I also changed the implementation to set the dirSeek after coming across the node the first time i.e. cleaning up all files in dir. Next iteration will start iterating through the sub dirs & add sub dir to stack one by one. One iteration at ta time.

Reply:
This looks great. Thanks for the detailed suggestion.